### ISSUE TITLE
[BUG] Missed null checks in record constructor when non-Lombok annotations are used

### ISSUE THREAD

**Issue by a--v--k** [projectlombok/lombok#3743]
**Describe the bug**

Lombok does not generate record constructor with nullability checks for a record with annotated fields - if non-Lombok annotations are used for. The following annotations are affected:
- Google Findbugs annotations
- Jakarta validation annotations
- Jetbrains annotations
- Spring annotations

When same annotations are used on fields of data class (annotated with @ Data and @ Builder) - constructor with nullability checks is generated properly.

**To Reproduce**
Please see the following [test project](https://github.com/a--v--k/lombok-issue-20240910). It contains all required annotated classes and unit tests.

**Expected behavior**

Lombok should generate record constructor with nullability checks for a record with annotated fields  even if non-Lombok annotations are used for.

**Version info (please complete the following information):**
 - Lombok version - 1.18.34
 - Lombok gradle plugin version - 8.10
 - Eclipse Temurin JDK 21.0.4
 - Gradle 8.9

**Additional context**

**Comment by GregDThomas** [comment]
Came here to raise this exact same issue :) 

Did get as far as writing a smaller test case for it, so will share that in case it's useful.

```java
    import org.springframework.lang.NonNull;

    @Test
    void checkNonNull() {

        assertThatThrownBy(() -> TestClass.builder().build())
            .describedAs("Checking class has non-null fields")
            // passes
            .isInstanceOf(NullPointerException.class);

        assertThatThrownBy(() -> TestRecord.builder().build())
            .describedAs("Checking record has non-null fields")
            // fails with Expecting code to raise a throwable.  
            .isInstanceOf(NullPointerException.class);

    }

    @Builder
    public static class TestClass {
        @org.springframework.lang.NonNull
        private final String value;
    }

    @Builder
    record TestRecord(@org.springframework.lang.NonNull String value) {

    }
```

**Comment by masiljangajji** [comment]
In the case of record, the compact constructor is automatically generated, and there is no separate handling for it, which makes it seem that null checks cannot be performed. However, with Lombok, it explicitly handles the compact constructor, allowing for null checks.

**Comment by GregDThomas** [comment]
> In the case of record, the compact constructor is automatically generated

Maybe, but the value of the fields could be checked when using the builder in the build() method.

**Comment by masiljangajji** [comment]
> Maybe, but the value of the fields could be checked when using the builder in the build() method.

In the case of a class, the @NonNull annotation is applied to the constructor generated by Lombok's @Builder, ensuring that the field is checked for null during object creation.

However, for a record, Java internally generates a compact constructor, and Lombok's @Builder respects this by not creating a new constructor. Instead, it uses the compact constructor generated by Java.

As a result, when you call TestRecord.builder().build(), a NullPointerException (NPE) does not occur, as the @NonNull annotation does not modify the behavior of the compact constructor.

However, in the case of Lombok's @NonNull, it introduces additional steps by creating a modifiable constructor instead of relying on the record's compact constructor. As a result, an NPE is triggered.

**Comment by GregDThomas** [comment]
> However, in the case of Lombok's @nonnull, it introduces additional steps by creating a modifiable constructor instead of relying on the record's compact constructor. As a result, an NPE is triggered.

Yes, but given that Lombok generally respects "other" nonnull annotations, why shouldn't it do so here?

**Comment by masiljangajji** [comment]
can you explain it in more detail?

**Comment by masiljangajji** [comment]
> 명시적으로 언급되지는 않았지만 이는 일반 클래스의 생성자/빌더에서도 발생합니다. 하지만 Java 레코드는 아닙니다.

Ah, I see. However, it seems like only the person who actually developed the feature would know such details.

**Comment by Rawi01** [comment]
It only happens if you generate your constructor/getter/setter/etc. with lombok, not if you use the record standard constructor. Lombok usually only modifies or generates things when you tell it to do so by adding an annotation. 

I think it would be quite confusing if lombok randomly modifies other records without any lombok annotation. How would someone without in-depth knowledge of lombok understand this?

**Comment by GregDThomas** [comment]
Lombok already does this if tagged `@lombok.NonNull`. The current lack of inconsistency is IMHO extremely confusing. Consider the following four essentially identical classes/records:
```java
    @lombok.Builder
    public static class TestLombokClass {
        @lombok.NonNull
        private final String value;
    }

    @lombok.Builder
    public static class TestSpringClass {
        @org.springframework.lang.NonNull
        private final String value;
    }

    @lombok.Builder
    record TestLombokRecord(@lombok.NonNull String value) {
    }

    @lombok.Builder
    record TestSpringRecord(@org.springframework.lang.NonNull String value) {
    }
```
At attempt to invoke 
* `TestLombokClass.builder.build()` fails, because value is null.
* `TestSpringClass.builder.build()` fails, because value is null.
* `TestLombokRecord.builder.build()` fails, because value is null.

An attempt to invoke
* `TestSpringRecord.builder().build()` succeeds, creating a record with a `null` value.

**Comment by a--v--k** [comment]
> I think it would be quite confusing if lombok randomly modifies other records without any lombok annotation. How would someone without in-depth knowledge of lombok understand this?

First of all, lombok create constructors with nullability checks for `@Data` classes even if non-lombok annotations are used on fields/constructor parameter (excepting jakarta annotations - why?). But Lombok modifies record constructors only if Lombok annotations are used.

In the test project I show the difference ( `[+]` means that lombok creates/modifies constructor, `[-]` that not):

| Annotations           | Data class                                                                                                                                    | Record                                                                                                                                  | 
|-----------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------|
| Findbug               | [**DataClassWithFindbugsAnnotations  [+]**](src/main/java/org/avk/lombok/issue/annotated/by/findbugs/DataClassWithFindbugsAnnotations.java)   | [**RecordWithFindbugsAnnotations  [-]**](src/main/java/org/avk/lombok/issue/annotated/by/findbugs/RecordWithFindbugsAnnotations.java)   |
| Jakarta validation    | [**DataClassWithJakartaAnnotations   [-]**](src/main/java/org/avk/lombok/issue/annotated/by/jakarta/DataClassWithJakartaAnnotations.java)     | [**RecordWithJakartaAnnotations   [-]**](src/main/java/org/avk/lombok/issue/annotated/by/jakarta/RecordWithJakartaAnnotations.java)     |
| Jetbrains annotations | [**DataClassWithJetbrainsAnnotations [+]**](src/main/java/org/avk/lombok/issue/annotated/by/jetbrains/DataClassWithJetbrainsAnnotations.java) | [**RecordWithJetbrainsAnnotations [-]**](src/main/java/org/avk/lombok/issue/annotated/by/jetbrains/RecordWithJetbrainsAnnotations.java) |
| Lombok                | [**DataClassWithLombokAnnotations    [+]**](src/main/java/org/avk/lombok/issue/annotated/by/lombok/DataClassWithLombokAnnotations.java)       | [**RecordWithLombokAnnotations    [+]**](src/main/java/org/avk/lombok/issue/annotated/by/lombok/RecordWithLombokAnnotations.java)       |
| Spring                | [**DataClassWithSpringAnnotations    [+]**](src/main/java/org/avk/lombok/issue/annotated/by/spring/DataClassWithSpringAnnotations.java)       | [**RecordWithSpringAnnotations    [-]**](src/main/java/org/avk/lombok/issue/annotated/by/spring/RecordWithSpringAnnotations.java)       |

**it would be great if team will add this ability (to modify record constructor according to field/parameter annotation) with a new configuration option (implemented as boolean flag or as list of annotation libraries for which this feature should be switched on).**

**Comment by GregDThomas** [comment]
> A possible solution would be to allow @AllArgsConstructor on records, then it becomes clear that the constructor is generade by lombok.

Yes, that also works - especially if it includes the implicit  @AllArgsConstructor introduced e.g. by the @Builder
